### PR TITLE
DEV: Unify code between PdfReader and PdfWriter

### PR DESCRIPTION
**Pull request by pubpub-zz**
requires #2495 
closes #2531

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma  / @stefan6419846 
I've started to prepare commonilisation but I've met a discrepancy in return types from pdf_header between PdfReader and PdfWriter.
I consider that this is acceptable to align PdfWriter onto PdfReader. Your opinions?

**Comment by stefan6419846** [comment]
> I consider that this is acceptable to align PdfWriter onto PdfReader. Your opinions?

How likely do you consider this to be an issue on the user side, id est how likely is it that this change is going to break user code? We do not have a real policy for these cases (where a deprecation process is not really feasible) at the moment and I do not see an easy solution for it, but I do not like knowingly breaking user code without a major release either, especially when we have a deprecation policy.

**Comment by pubpub-zz** [comment]
@stefan6419846  / @MartinThoma 
still progressing on my dev, facing new questions 😉
Although we have now a good test bank, I'm wondering if we should not  propose a first beta release. Do you have any idea how we could propose that ?

**Comment by stefan6419846** [comment]
> Although we have now a good test bank, I'm wondering if we should not propose a first beta release. Do you have any idea how we could propose that ?

In theory there a pre-releases on GitHub itself and https://test.pypi.org, but this has been uncommon for *pypdf* until now and I suspect that general attention/adaption will be rather limited.

**Comment by MartinThoma** [comment]
I haven't seen test.pypi.org being used.

I did see pre-releases (e.g. https://pypi.org/project/eth-tester/0.1.0b7 or https://pypi.org/project/pydantic/2.6.0b1/ )

**Comment by stefan6419846** [comment]
@pubpub-zz Are you okay with marking this PR as WIP/draft for now and unmarking it once its ready for actual review? Otherwise, I am not always sure whether reviewing makes sense from my side or not as you are changing anything about this anyway.

**Comment by pubpub-zz** [comment]
@stefan6419846 
even if all new lines are not covered, The refactoring has improved the coverage reaching 95.62% 
Documentation should be good : I've left the inherited functions in PdfReader and PdfWriter documentation (PdfDocCommon should remain transparent to users)

File: docs/index.rst [hunk 1]
```diff
@@ -47,6 +47,7 @@ You can contribute to `pypdf on GitHub <https://github.com/py-pdf/pypdf>`_.
    :caption: API Reference
    :maxdepth: 1
 
+   modules/PdfDocCommon
```
Comment (stefan6419846) [inline docs/index.rst:50]: Do we really consider this the top-level class already?

File: docs/modules/generic.rst [hunk 2]
```diff
@@ -7,12 +7,19 @@ Generic PDF objects
...
     :exclude-members: Destination, Field, Fit, RectangleObject
 
+
 .. autoclass:: pypdf._protocols.PdfObjectProtocol
     :members:
     :undoc-members:
     :show-inheritance:
 
 
+.. autoclass:: pypdf._protocols.PdfCommonDocProtocol
```
Comment (stefan6419846) [inline docs/modules/generic.rst:17]: We might want to move the protocols to the new `PdfDocCommon` page, or do you think that it does not belong there?

File: pypdf/_doc_common.py [hunk 3]
```diff
@@ -0,0 +1,1303 @@
+# Copyright (c) 2006, Mathieu Fenniak
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1]: Do we really need this header? As far as I can see, we have no consistent license headers, thus it should not really be required unless you want to explicitly declare your copyright here.

File: pypdf/_doc_common.py [hunk 4]
```diff
@@ -0,0 +1,1303 @@
...
+        is the offset from UTC.
+        """
+        return self.get(DI.MOD_DATE)
+
+
+class PdfDocCommon:
+    """
+    Common functions from PdfWriter and PdfReader objects.
+
+    This root class is strongly abstracted
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:250]:
```suggestion
    This root class is strongly abstracted.
```

File: pypdf/_doc_common.py [hunk 5]
```diff
@@ -0,0 +1,1303 @@
...
+        include: Union[None, Iterable[str]] = None,
+        exclude: Union[None, Iterable[str]] = None,
+    ) -> Dict[str, Any]:
+        """
+
+    # ????#@property
+    # ????#def metadata(self) -> Optional[DocumentInformation]:
+
+    # ????#@property
+    # ????#def xmp_metadata(self) -> Optional[XmpInformation]:
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:320]: The previous line should probably be cleaned up or discussed further.

File: pypdf/_doc_common.py [hunk 6]
```diff
@@ -0,0 +1,1303 @@
...
+
+    flattened_pages: Optional[List[PageObject]] = None
+
+    ### common
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:324]: Do we really need these comments?

File: pypdf/_doc_common.py [hunk 7]
```diff
@@ -0,0 +1,1303 @@
...
+    ## common
+    @property
+    def named_destinations(self) -> Dict[str, Any]:
+        """
+        A read-only dictionary which maps names to
+        :class:`Destinations<pypdf.generic.Destination>`
+        """
+        return self._get_named_destinations()
+
+    # ???? add get_named_dest_root caution if
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:374]: Incomplete obsolete comment?

File: pypdf/_doc_common.py [hunk 8]
```diff
@@ -0,0 +1,1303 @@
...
+                        val = val["/D"].get_object()
+                    else:
+                        continue
+                dest = self._build_destination(k__, val)
+                if dest is not None:
+                    retval[k__] = dest
+        return retval
+
+    # A select group of relevant field attributes. For the complete list,
+    # see section 8.6.2 of the PDF 1.7 reference.
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:448]: What does this refer to?

File: pypdf/_doc_common.py [hunk 9]
```diff
@@ -0,0 +1,1303 @@
...
+
+        if "/Fields" in tree:
+            fields = cast(ArrayObject, tree["/Fields"])
+            for f in fields:
+                field = f.get_object()
+                self._build_field(field, retval, fileobj, field_attributes)
+
+        return retval
+
+    #### ??? : a mettre en PR indep
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:500]: We probably do not want to keep French TODO comments here.

File: pypdf/_doc_common.py [hunk 10]
```diff
@@ -0,0 +1,1303 @@
...
+        It's an array of dictionaries with "/F" and "/I" properties or
+        None if there are no articles.
+        """
+        catalog = self.root_object
+        if CO.THREADS in catalog:
+            return cast("ArrayObject", catalog[CO.THREADS])
+        else:
+            return None
+
+    def _get_num_pages(self) -> int:
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:823]: Do we really need `get_num_pages` and `_get_num_pages`?

File: pypdf/_doc_common.py [hunk 11]
```diff
@@ -0,0 +1,1303 @@
...
+        # the PDF file's page count is used in this case. Otherwise,
+        # the original method (flattened page count) is used.
+        if self.is_encrypted:
+            return self.root_object["/Pages"]["/Count"]  # type: ignore
+        else:
+            if self.flattened_pages is None:
+                self._flatten()
+            return len(self.flattened_pages)  # type: ignore
+
+    # ????# to be merged ?
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:844]: TODO comment?

File: pypdf/_doc_common.py [hunk 12]
```diff
@@ -0,0 +1,1303 @@
...
+            "assemble": UserAccessPermissions.ASSEMBLE_DOC,
+            "print_high_quality": UserAccessPermissions.PRINT_TO_REPRESENTATION,
+        }
+
+        return {
+            key: permissions_code & flag != 0
+            for key, flag in permissions_mapping.items()
+        }
+
+    # ????#
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1160]: TODO comment?

File: pypdf/_reader.py [hunk 13]
```diff
@@ -479,218 +270,6 @@ def _get_page(self, page_number: int) -> PageObject:
...
-        ff = {}
-        for field, value in formfields.items():
-            if value.get("/FT") == "/Tx":
-                if full_qualified_name:
-                    ff[field] = value.get("/V")
-                else:
-                    ff[indexed_key(cast(str, value["/T"]), ff)] = value.get("/V")
-        return ff
-
     def get_pages_showing_field(
```
Comment (stefan6419846) [inline pypdf/_reader.py:273]: We cannot use the common implementation here?

File: pypdf/_reader.py [hunk 14]
```diff
@@ -1423,6 +608,16 @@ def cache_indirect_object(
             obj.indirect_reference = IndirectObject(idnum, generation, self)
         return obj
 
+    def _replace_object(self, indirect: IndirectObject, obj: PdfObject) -> PdfObject:
+        # function reserved for future dev
+        if indirect.pdf != self:
+            raise ValueError("Can not update pdfreader with external object")
```
Comment (stefan6419846) [inline pypdf/_reader.py:614]:
```suggestion
            raise ValueError("Cannot update PdfReader with external object")
```

File: pypdf/_reader.py [hunk 15]
```diff
@@ -1423,6 +608,16 @@ def cache_indirect_object(
...
+        if (indirect.generation, indirect.idnum) not in self.resolved_objects:
+            raise ValueError("Can not find referenced object")
```
Comment (stefan6419846) [inline pypdf/_reader.py:616]:
```suggestion
            raise ValueError("Cannot find referenced object")
```

File: pypdf/_writer.py [hunk 16]
```diff
@@ -311,17 +361,73 @@ def _add_page(
...
+
+            clean: replace PageObject with NullObject to prevent destination,
+                annotation to reference a detached page
+        """
+        if self.flattened_pages is None:
+            return
+        if isinstance(page, IndirectObject):
+            p = page.get_object()
+            if not isinstance(p, PageObject):
+                logger_warning("IndirectObject is not referencing to a page", __name__)
```
Comment (stefan6419846) [inline pypdf/_writer.py:399]:
```suggestion
                logger_warning("IndirectObject is not referencing a page", __name__)
```

File: pypdf/_writer.py [hunk 17]
```diff
@@ -311,17 +361,73 @@ def _add_page(
...
+                return
+            page = p
+        if not isinstance(page, int):
+            try:
+                page = self.flattened_pages.index(page)
+            except ValueError:
+                logger_warning("Can't find page in pages", __name__)
```
Comment (stefan6419846) [inline pypdf/_writer.py:406]:
```suggestion
                logger_warning("Cannot find page in pages", __name__)
```

File: pypdf/_writer.py [hunk 18]
```diff
@@ -311,17 +361,73 @@ def _add_page(
...
+                return
+        if not (0 <= page < len(self.flattened_pages)):
+            logger_warning("page number is out of range", __name__)
```
Comment (stefan6419846) [inline pypdf/_writer.py:409]:
```suggestion
            logger_warning("Page number is out of range", __name__)
```

File: pypdf/_writer.py [hunk 19]
```diff
@@ -535,29 +591,7 @@ def insert_blank_page(
...
-            try:
-                page, typ = oa[0:2]
-                array = oa[2:]
-                fit = Fit(typ, tuple(array))
-                return Destination("OpenAction", page, fit)
-            except Exception as exc:
-                raise Exception(f"Invalid Destination {oa}: {exc}")
-        else:
-            return None
+        return super().open_destination
```
Comment (stefan6419846) [inline pypdf/_writer.py:594]: Do we really need to overwrite this method when we only have the `super` call anyway?

File: tests/scripts/test_make_release.py [hunk 20]
```diff
@@ -20,58 +20,60 @@
 def test_get_git_commits_since_tag():
     make_release = pytest.importorskip("make_release")
 
-    with open(COMMITS__VERSION_4_0_1, mode="rb") as commits, \
-            mock.patch("urllib.request.urlopen", side_effect=lambda n: commits), \
-            mock.patch("subprocess.check_output", return_value=GIT_LOG__VERSION_4_0_1):
+    with open(COMMITS__VERSION_4_0_1, mode="rb") as commits, mock.patch(
```
Comment (stefan6419846) [inline tests/scripts/test_make_release.py:23]: These changes from *black* or a more recent *ruff* are unrelated and should be reverted.

File: tests/test_page.py [hunk 21]
```diff
@@ -1203,8 +1203,11 @@ def test_merge_transformed_page_into_blank():
     inserted_blank = writer.add_page(blank)
     assert blank.page_number is None  # the inserted page is a clone
     assert inserted_blank.page_number == len(writer.pages) - 1
-    del writer._pages.get_object()["/Kids"][-1]
+    writer.remove_page(inserted_blank.indirect_reference)
```
Comment (stefan6419846) [inline tests/test_page.py:1206]: As this has been an internal attribute (`_pages`), we do not consider this a breaking change?

**Comment by stefan6419846** [comment]
Thanks. I have added some remarks about lines I am not sure about or which might need further fixes.

>  even if all new lines are not covered, The refactoring has improved the coverage reaching 95.62%

Some of this might be related to the coverage pragma as well, while others might indeed could be fixed (like trying to decrypt a not encrypted PDF file).

File: docs/index.rst (hunk 1 above)
Comment (pubpub-zz) [reply docs/index.rst:50]: can you clarify?

File: docs/modules/generic.rst (hunk 2 above)
Comment (pubpub-zz) [reply docs/modules/generic.rst:17]: the protocols are for me of no interest : we have to refer to the implementations PdfReader/PdfWriter ; even when  PdfCommonDoc  is called it shall be considered to use a PdfReader or PdfWriter

File: pypdf/_doc_common.py (hunk 3 above)
Comment (pubpub-zz) [reply pypdf/_doc_common.py:1]: _doc_common.py is mainly a copy/paste/extract of _reader.py and _writer.py. Not sure we should remove the copyright

File: pypdf/_doc_common.py (hunk 8 above)
Comment (pubpub-zz) [reply pypdf/_doc_common.py:448]: this comment used to be there (in _reader.py  or _writer.py) . unchanged

File: pypdf/_doc_common.py (hunk 10 above)
Comment (pubpub-zz) [reply pypdf/_doc_common.py:823]: functions have been kept unchanged

File: docs/index.rst (hunk 1 above)
Comment (stefan6419846) [reply docs/index.rst:50]: In the index, we currently start with the PdfReader and PdfWriter class, then with the helper classes/modules. Should we really put PdfDocCommon at the top as the "most important" class?

File: docs/modules/generic.rst (hunk 2 above)
Comment (stefan6419846) [reply docs/modules/generic.rst:17]: These are required for the cross-references. I am just not sure whether they might be moved to another documentation file/page.

File: pypdf/_doc_common.py (hunk 3 above)
Comment (stefan6419846) [reply pypdf/_doc_common.py:1]: Thanks for the explanation. It does not really hurt, thus just keep it here.

File: pypdf/_doc_common.py (hunk 10 above)
Comment (stefan6419846) [reply pypdf/_doc_common.py:823]: But we cannot we drop `_get_num_pages` and replace it with `get_num_pages`? `_get_num_pages` is internal and thus does not require a deprecation process. Keeping code duplicates does not feel right.

File: docs/index.rst (hunk 1 above)
Comment (pubpub-zz) [reply docs/index.rst:50]: Don't think so: I consider this class as an abstract one which is just here to centralized some code. This is why the help page should not describe the members and why the inherited functions are detailed in pdfreader and pdfwriter, proving duplicated text

Comment (stefan6419846) [reply docs/index.rst:50]: Then this line needs to be changed to move PdfDocCommon a bit more to the bottom.

File: tests/test_page.py (hunk 21 above)
Comment (pubpub-zz) [reply tests/test_page.py:1206]: this hack for test coverage has been rewritten. remove_page is a slight extension of the existing 'del writer.pages[int]` allowing to pass a `PageObject` or an `IndirectObject`

File: tests/scripts/test_make_release.py (hunk 20 above)
Comment (pubpub-zz) [reply tests/scripts/test_make_release.py:23]: this is due to main merging. It should clean up during final merge

File: pypdf/_writer.py (hunk 19 above)
Comment (pubpub-zz) [reply pypdf/_writer.py:573]: Yes without this approach I've not been able to add the setter function (applicable only to PdfWriter)

File: pypdf/_reader.py (hunk 13 above)
Comment (pubpub-zz) [reply pypdf/_reader.py:262]: fixed duplication

File: pypdf/_doc_common.py (hunk 10 above)
Comment (pubpub-zz) [reply pypdf/_doc_common.py:862]: agreed fixed

**Comment by pubpub-zz** [comment]
@stefan6419846 
I propose the PR with the current coverage. improvement could be identified later

File: pypdf/_doc_common.py [hunk 22]
```diff
@@ -0,0 +1,1363 @@
...
+        Calculate the number of pages in this PDF file.
+
+        Returns:
+            The number of pages of the parsed PDF file
+
+        Raises:
+            PdfReadError: if file is encrypted and restrictions prevent
+                this action.
+        """
+        # Flattened pages will not work on an Encrypted PDF;
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:346]:
```suggestion
        # Flattened pages will not work on an encrypted PDF;
```

File: pypdf/_doc_common.py [hunk 23]
```diff
@@ -0,0 +1,1363 @@
...
+        else:
+            if self.flattened_pages is None:
+                self._flatten()
+            assert self.flattened_pages is not None
+            return len(self.flattened_pages)
+
+    def get_page(self, page_number: int) -> PageObject:
+        """
+        Retrieve a page by number from this PDF file.
+        Most of the time ```.pages[page_number]``` is prefered
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:360]:
```suggestion
        Most of the time ```.pages[page_number]``` is preferred.
```

File: pypdf/_doc_common.py [hunk 24]
```diff
@@ -0,0 +1,1363 @@
...
+                i += 1
+                if isinstance(value, DictionaryObject):
+                    if "/D" in value:
+                        value = value["/D"]
+                    else:
+                        continue
+                dest = self._build_destination(key, value)  # type: ignore
+                if dest is not None:
+                    retval[key] = dest
+        else:  # case where Dests is in root catalog (PDF 1.7 specs, §2 about PDF1.1
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:474]:
```suggestion
        else:  # case where Dests is in root catalog (PDF 1.7 specs, §2 about PDF 1.1)
```

File: pypdf/_doc_common.py [hunk 25]
```diff
@@ -0,0 +1,1363 @@
...
+            outline_item.indirect_reference = node.indirect_reference
+        except AttributeError:
+            pass
+        return outline_item
+
+    @property
+    def pages(self) -> List[PageObject]:
+        """
+        Property that emulates a list of :class:`PageObject<pypdf._page.PageObject>`.
+        this property allows to get a page or  a range of pages.
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1004]:
```suggestion
        this property allows to get a page or a range of pages.
```

File: pypdf/_doc_common.py [hunk 26]
```diff
@@ -0,0 +1,1363 @@
...
+
+        For PdfWriter Only:
+        It provides also capability to remove a page/range of page from the list
+        (through del operator)
+        Note: only the page entry is removed. As the objects beneath can be used
+        somewhere else.
+        a solution to completely remove them - if they are not used somewhere -
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1011]:
```suggestion
        A solution to completely remove them - if they are not used anywhere -
```

File: pypdf/_doc_common.py [hunk 27]
```diff
@@ -0,0 +1,1363 @@
...
+        is to write to a buffer/temporary and to then load it into a new PdfWriter
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1012]:
```suggestion
        is to write to a buffer/temporary file and to load it into a new PdfWriter
```

File: pypdf/_doc_common.py [hunk 28]
```diff
@@ -0,0 +1,1363 @@
...
+        object.
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1013]:
```suggestion
        object afterwards.
```

File: pypdf/_doc_common.py [hunk 29]
```diff
@@ -0,0 +1,1363 @@
...
+        Remove page from pages list.
+
+        Args:
+            page: int / PageObject / IndirectObject
+                PageObject : page to be removed. If the page appears many times
+                only the first one will be removed
+
+                IndirectObject: Reference to page to be removed
+
+                int : page number to be removed
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1150]:
```suggestion
                int: Page number to be removed
```

File: pypdf/_doc_common.py [hunk 30]
```diff
@@ -0,0 +1,1363 @@
...
+
+        Args:
+            filename: If filename is None, then a dictionary of all attachments
+                will be returned, where the key is the filename and the value
+                is the content. Otherwise, a dictionary with just a single key
+                - the filename - and its content will be returned.
+
+        Returns:
+            dictionary of filename -> Union[bytestring or List[ByteString]]
+            if the filename exists multiple times a List of the different version will be provided
```
Comment (stefan6419846) [inline pypdf/_doc_common.py:1315]:
```suggestion
            If the filename exists multiple times a list of the different versions will be provided.
```

File: pypdf/_reader.py [hunk 31]
```diff
@@ -356,7 +167,15 @@ def _info(self) -> Optional[DictionaryObject]:
...
         """
         info = self.trailer.get(TK.INFO, None)
-        return None if info is None else cast(DictionaryObject, info.get_object())
+        if info is None:
+            return None
+        else:
+            info = info.get_object()
+            if info is None:
+                raise PdfReadError(
+                    "trailer not found or does not point to document information directory"
```
Comment (stefan6419846) [inline pypdf/_reader.py:176]:
```suggestion
                    "Trailer not found or does not point to document information directory"
```

File: pypdf/_reader.py [hunk 32]
```diff
@@ -1222,30 +329,19 @@ def _get_object_from_stream(
                     __name__,
                 )
 
-                if self.strict:
-                    raise PdfReadError(f"Can't read object stream: {exc}")
+                if self.strict:  # pragma: no cover
+                    raise PdfReadError(
+                        f"Can't read object stream: {exc}"
```
Comment (stefan6419846) [inline pypdf/_reader.py:334]:
```suggestion
                        f"Cannot read object stream: {exc}"
```

**Comment by pubpub-zz** [comment]
@stefan6419846 
are you expecting any further work before merging?

**Review by stefan6419846 — APPROVED** [review]

**Comment by stefan6419846** [comment]
@pubpub-zz No, I just did not have enough time this morning to finish this regarding incorporating the latest changes from *main*.

**Pull request merged**